### PR TITLE
feat: add example stubs

### PR DESCRIPTION
**Pull request by bdura**
## Description

This PR adds a stubs file for `spacy.training.example`. It also fixes a few typing-related issues.

### Types of change

This is an enhancement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by bdura** [comment]
I only added evident type hints. I'll add those later on, either in this PR or the next - let me know what works best since I'll likely have little time for that in the next few days.

File: spacy/tokens/doc.pyi [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    Union,
+    overload,
+)
```
Comment (adrianeboyd) [inline spacy/tokens/doc.pyi:15]:
```suggestion
from typing import Any, Callable, Dict, Iterable, Iterator, List, Optional
from typing import Protocol, Sequence, Tuple, Union, overload
from pathlib import Path
```

File: spacy/tokens/doc.pyi [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
+
+import numpy as np
```
Comment (adrianeboyd) [inline spacy/tokens/doc.pyi:17]:
```suggestion
import numpy
```

File: spacy/tokens/doc.pyi (hunk 1 above)
Comment (adrianeboyd) [reply spacy/tokens/doc.pyi:15]: Let me just do this directly...

File: spacy/training/corpus.py [hunk 3]
```diff
@@ -19,14 +20,19 @@
 FILE_TYPE = ".spacy"
 
 
+class ReaderProtocol(Protocol):
+    def __call__(self, nlp: "Language") -> Iterable[Example]:
```
Comment (adrianeboyd) [inline spacy/training/corpus.py:24]: I thought `Iterator[Example]` was correct throughout here?

Comment (svlandeg) [reply spacy/training/corpus.py:24]: Answered [here](https://github.com/explosion/spaCy/pull/12801#pullrequestreview-1518042573).

**Pull request merged**